{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "04000c5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re\n",
    "import os\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e5f76f41",
   "metadata": {},
   "outputs": [],
   "source": [
    "from os import path\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "03a50eac",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.sparse import csc_matrix, csr_matrix, save_npz\n",
    "from scipy.io import mmread, mmwrite"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a9bad0d",
   "metadata": {},
   "source": [
    "# Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9d6e2b71",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/gpfs/gibbs/pi/zhao/yw599/Multiome/data/BMMC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c6dd462a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('GSE194122_openproblems_neurips2021_multiome_BMMC_processed.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "51b0d2f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 69249 × 129921\n",
       "    obs: 'GEX_pct_counts_mt', 'GEX_n_counts', 'GEX_n_genes', 'GEX_size_factors', 'GEX_phase', 'ATAC_nCount_peaks', 'ATAC_atac_fragments', 'ATAC_reads_in_peaks_frac', 'ATAC_blacklist_fraction', 'ATAC_nucleosome_signal', 'cell_type', 'batch', 'ATAC_pseudotime_order', 'GEX_pseudotime_order', 'Samplename', 'Site', 'DonorNumber', 'Modality', 'VendorLot', 'DonorID', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'QCMeds', 'DonorSmoker'\n",
       "    var: 'feature_types', 'gene_id'\n",
       "    uns: 'ATAC_gene_activity_var_names', 'dataset_id', 'genome', 'organism'\n",
       "    obsm: 'ATAC_gene_activity', 'ATAC_lsi_full', 'ATAC_lsi_red', 'ATAC_umap', 'GEX_X_pca', 'GEX_X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3483b4c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CD8+ T                 11589\n",
       "CD14+ Mono             10843\n",
       "NK                      6929\n",
       "CD4+ T activated        5526\n",
       "Naive CD20+ B           5052\n",
       "Erythroblast            4916\n",
       "CD4+ T naive            4398\n",
       "Transitional B          2810\n",
       "Proerythroblast         2300\n",
       "CD16+ Mono              1894\n",
       "B1 B                    1890\n",
       "Normoblast              1780\n",
       "Lymph prog              1779\n",
       "G/M prog                1203\n",
       "pDC                     1191\n",
       "HSC                     1072\n",
       "CD8+ T naive            1012\n",
       "MK/E prog                884\n",
       "cDC2                     859\n",
       "ILC                      835\n",
       "Plasma cell              379\n",
       "ID2-hi myeloid prog      108\n",
       "Name: cell_type, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.cell_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c5cb909f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Separate ATAC and RNA data\n",
    "is_gex_col = adata.var.feature_types == 'GEX'\n",
    "adata_rna = adata[:, is_gex_col]\n",
    "adata_atac = adata[:, ~is_gex_col]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7784be15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 69249 × 13431\n",
       "    obs: 'GEX_pct_counts_mt', 'GEX_n_counts', 'GEX_n_genes', 'GEX_size_factors', 'GEX_phase', 'ATAC_nCount_peaks', 'ATAC_atac_fragments', 'ATAC_reads_in_peaks_frac', 'ATAC_blacklist_fraction', 'ATAC_nucleosome_signal', 'cell_type', 'batch', 'ATAC_pseudotime_order', 'GEX_pseudotime_order', 'Samplename', 'Site', 'DonorNumber', 'Modality', 'VendorLot', 'DonorID', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'QCMeds', 'DonorSmoker'\n",
       "    var: 'feature_types', 'gene_id'\n",
       "    uns: 'ATAC_gene_activity_var_names', 'dataset_id', 'genome', 'organism'\n",
       "    obsm: 'ATAC_gene_activity', 'ATAC_lsi_full', 'ATAC_lsi_red', 'ATAC_umap', 'GEX_X_pca', 'GEX_X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a8033c78",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 69249 × 116490\n",
       "    obs: 'GEX_pct_counts_mt', 'GEX_n_counts', 'GEX_n_genes', 'GEX_size_factors', 'GEX_phase', 'ATAC_nCount_peaks', 'ATAC_atac_fragments', 'ATAC_reads_in_peaks_frac', 'ATAC_blacklist_fraction', 'ATAC_nucleosome_signal', 'cell_type', 'batch', 'ATAC_pseudotime_order', 'GEX_pseudotime_order', 'Samplename', 'Site', 'DonorNumber', 'Modality', 'VendorLot', 'DonorID', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'QCMeds', 'DonorSmoker'\n",
       "    var: 'feature_types', 'gene_id'\n",
       "    uns: 'ATAC_gene_activity_var_names', 'dataset_id', 'genome', 'organism'\n",
       "    obsm: 'ATAC_gene_activity', 'ATAC_lsi_full', 'ATAC_lsi_red', 'ATAC_umap', 'GEX_X_pca', 'GEX_X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "586b4f5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get single RNA data\n",
    "batch_rna = ['s1d2', 's1d3', 's3d3', 's4d9', 's3d10']\n",
    "adata_rna_single = adata_rna[adata_rna.obs.batch.isin(batch_rna)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "086b4e0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 26450 × 13431\n",
       "    obs: 'GEX_pct_counts_mt', 'GEX_n_counts', 'GEX_n_genes', 'GEX_size_factors', 'GEX_phase', 'ATAC_nCount_peaks', 'ATAC_atac_fragments', 'ATAC_reads_in_peaks_frac', 'ATAC_blacklist_fraction', 'ATAC_nucleosome_signal', 'cell_type', 'batch', 'ATAC_pseudotime_order', 'GEX_pseudotime_order', 'Samplename', 'Site', 'DonorNumber', 'Modality', 'VendorLot', 'DonorID', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'QCMeds', 'DonorSmoker'\n",
       "    var: 'feature_types', 'gene_id'\n",
       "    uns: 'ATAC_gene_activity_var_names', 'dataset_id', 'genome', 'organism'\n",
       "    obsm: 'ATAC_gene_activity', 'ATAC_lsi_full', 'ATAC_lsi_red', 'ATAC_umap', 'GEX_X_pca', 'GEX_X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_rna_single"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "89a747ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get single ATAC data\n",
    "batch_atac = ['s2d4', 's2d5', 's3d7', 's4d8', 's3d6']\n",
    "adata_atac_single = adata_atac[adata_atac.obs.batch.isin(batch_atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "49edc817",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 24332 × 116490\n",
       "    obs: 'GEX_pct_counts_mt', 'GEX_n_counts', 'GEX_n_genes', 'GEX_size_factors', 'GEX_phase', 'ATAC_nCount_peaks', 'ATAC_atac_fragments', 'ATAC_reads_in_peaks_frac', 'ATAC_blacklist_fraction', 'ATAC_nucleosome_signal', 'cell_type', 'batch', 'ATAC_pseudotime_order', 'GEX_pseudotime_order', 'Samplename', 'Site', 'DonorNumber', 'Modality', 'VendorLot', 'DonorID', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'QCMeds', 'DonorSmoker'\n",
       "    var: 'feature_types', 'gene_id'\n",
       "    uns: 'ATAC_gene_activity_var_names', 'dataset_id', 'genome', 'organism'\n",
       "    obsm: 'ATAC_gene_activity', 'ATAC_lsi_full', 'ATAC_lsi_red', 'ATAC_umap', 'GEX_X_pca', 'GEX_X_umap'\n",
       "    layers: 'counts'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac_single"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "6ce54c7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get multiome RNA data\n",
    "batch_rna = ['s1d1', 's2d1', 's4d1']\n",
    "adata_rna_mul = adata_rna[adata_rna.obs.batch.isin(batch_rna)]\n",
    "# Get multiome ATAC data\n",
    "batch_atac = ['s1d1', 's2d1', 's4d1']\n",
    "adata_atac_mul = adata_atac[adata_atac.obs.batch.isin(batch_atac)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "91d3d901",
   "metadata": {},
   "outputs": [],
   "source": [
    "# del unnecessary data\n",
    "del adata, adata_rna, adata_atac"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cba8126",
   "metadata": {},
   "source": [
    "## Save unimodal RNA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "840f3d92",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = '/gpfs/gibbs/pi/zhao/yw599/Multiome/data_pp/BMMC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4b22a475",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_counts = csr_matrix(adata_rna_single.layers['counts'])\n",
    "rna_features = adata_rna_single.var_names\n",
    "rna_cells = adata_rna_single.obs_names\n",
    "rna_annotation = np.array(adata_rna_single.obs.cell_type)\n",
    "rna_path = path.join(folder, 'RNA')\n",
    "if not os.path.exists(rna_path):\n",
    "    os.makedirs(rna_path)\n",
    "np.savetxt(path.join(rna_path, 'cells.txt'), rna_cells, fmt='%s')\n",
    "np.savetxt(path.join(rna_path, 'genes.txt'), rna_features, fmt='%s')\n",
    "np.savetxt(path.join(rna_path, 'annotations.txt'), rna_annotation, fmt='%s')\n",
    "mmwrite(path.join(rna_path, 'counts.mtx'), rna_counts, field = 'integer')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0eace43d",
   "metadata": {},
   "source": [
    "## Save unimodal ATAC files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "974f2864",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_counts = csr_matrix(adata_atac_single.layers['counts'])\n",
    "atac_features = adata_atac_single.var_names\n",
    "atac_cells = adata_atac_single.obs_names\n",
    "atac_annotation = np.array(adata_atac_single.obs.cell_type)\n",
    "atac_path = path.join(folder, 'ATAC')\n",
    "if not os.path.exists(atac_path):\n",
    "    os.makedirs(atac_path)\n",
    "np.savetxt(path.join(atac_path, 'cells.txt'), atac_cells, fmt='%s')\n",
    "np.savetxt(path.join(atac_path, 'genes.txt'), atac_features, fmt='%s')\n",
    "np.savetxt(path.join(atac_path, 'annotations.txt'), atac_annotation, fmt='%s')\n",
    "mmwrite(path.join(atac_path, 'counts.mtx'), atac_counts, field = 'integer')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a147a73e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# atac_counts = csr_matrix(adata_atac_single.obsm['ATAC_gene_activity'])\n",
    "# atac_features = adata_atac_single.uns['ATAC_gene_activity_var_names']\n",
    "# atac_cells = adata_atac_single.obs_names\n",
    "# atac_annotation = np.array(adata_atac_single.obs.cell_type)\n",
    "# atac_path = path.join(folder, 'ATAC.gene.activities')\n",
    "# np.savetxt(path.join(atac_path, 'cells.txt'), atac_cells, fmt='%s')\n",
    "# np.savetxt(path.join(atac_path, 'genes.txt'), atac_features, fmt='%s')\n",
    "# np.savetxt(path.join(atac_path, 'annotations.txt'), atac_annotation, fmt='%s')\n",
    "# mmwrite(path.join(atac_path, 'counts.mtx'), atac_counts, field = 'integer')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42b1e537",
   "metadata": {},
   "source": [
    "## Save multiome RNA files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9424b410",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_counts = csr_matrix(adata_rna_mul.layers['counts'])\n",
    "rna_features = adata_rna_mul.var_names\n",
    "rna_cells = adata_rna_mul.obs_names\n",
    "rna_annotation = np.array(adata_rna_mul.obs.cell_type)\n",
    "rna_path = path.join(folder, 'multiome/RNA')\n",
    "if not os.path.exists(rna_path):\n",
    "    os.makedirs(rna_path)\n",
    "np.savetxt(path.join(rna_path, 'cells.txt'), rna_cells, fmt='%s')\n",
    "np.savetxt(path.join(rna_path, 'genes.txt'), rna_features, fmt='%s')\n",
    "np.savetxt(path.join(rna_path, 'annotations.txt'), rna_annotation, fmt='%s')\n",
    "mmwrite(path.join(rna_path, 'counts.mtx'), rna_counts, field = 'integer')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc281283",
   "metadata": {},
   "source": [
    "## Save multiome ATAC files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "81c1f89b",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_counts = csr_matrix(adata_atac_mul.layers['counts'])\n",
    "atac_features = adata_atac_mul.var_names\n",
    "atac_cells = adata_atac_mul.obs_names\n",
    "atac_annotation = np.array(adata_atac_mul.obs.cell_type)\n",
    "atac_path = path.join(folder, 'multiome/ATAC')\n",
    "if not os.path.exists(atac_path):\n",
    "    os.makedirs(atac_path)\n",
    "np.savetxt(path.join(atac_path, 'cells.txt'), atac_cells, fmt='%s')\n",
    "np.savetxt(path.join(atac_path, 'genes.txt'), atac_features, fmt='%s')\n",
    "np.savetxt(path.join(atac_path, 'annotations.txt'), atac_annotation, fmt='%s')\n",
    "mmwrite(path.join(atac_path, 'counts.mtx'), atac_counts, field = 'integer')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ece8df5d",
   "metadata": {},
   "source": [
    "# Experimental design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1a8b6e49",
   "metadata": {},
   "outputs": [],
   "source": [
    "seed = 10"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee173460",
   "metadata": {},
   "source": [
    "## Subsample number of cells in RNA and ATAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0c525bb5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "0 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed0.txt\n",
      "26450 200\n",
      "0 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed0.txt\n",
      "26450 400\n",
      "0 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed0.txt\n",
      "26450 600\n",
      "0 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed0.txt\n",
      "26450 800\n",
      "0 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed0.txt\n",
      "26450 1000\n",
      "0 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed0.txt\n",
      "26450 3000\n",
      "0 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed0.txt\n",
      "26450 5000\n",
      "0 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed0.txt\n",
      "26450 10000\n",
      "0 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed0.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "0 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed0.txt\n",
      "22653 200\n",
      "0 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed0.txt\n",
      "22653 400\n",
      "0 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed0.txt\n",
      "22653 600\n",
      "0 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed0.txt\n",
      "22653 800\n",
      "0 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed0.txt\n",
      "22653 1000\n",
      "0 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed0.txt\n",
      "22653 3000\n",
      "0 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed0.txt\n",
      "22653 5000\n",
      "0 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed0.txt\n",
      "22653 10000\n",
      "0 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed0.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "1 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed1.txt\n",
      "26450 200\n",
      "1 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed1.txt\n",
      "26450 400\n",
      "1 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed1.txt\n",
      "26450 600\n",
      "1 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed1.txt\n",
      "26450 800\n",
      "1 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed1.txt\n",
      "26450 1000\n",
      "1 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed1.txt\n",
      "26450 3000\n",
      "1 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed1.txt\n",
      "26450 5000\n",
      "1 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed1.txt\n",
      "26450 10000\n",
      "1 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed1.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "1 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed1.txt\n",
      "22653 200\n",
      "1 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed1.txt\n",
      "22653 400\n",
      "1 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed1.txt\n",
      "22653 600\n",
      "1 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed1.txt\n",
      "22653 800\n",
      "1 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed1.txt\n",
      "22653 1000\n",
      "1 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed1.txt\n",
      "22653 3000\n",
      "1 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed1.txt\n",
      "22653 5000\n",
      "1 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed1.txt\n",
      "22653 10000\n",
      "1 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed1.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "2 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed2.txt\n",
      "26450 200\n",
      "2 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed2.txt\n",
      "26450 400\n",
      "2 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed2.txt\n",
      "26450 600\n",
      "2 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed2.txt\n",
      "26450 800\n",
      "2 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed2.txt\n",
      "26450 1000\n",
      "2 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed2.txt\n",
      "26450 3000\n",
      "2 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed2.txt\n",
      "26450 5000\n",
      "2 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed2.txt\n",
      "26450 10000\n",
      "2 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed2.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "2 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed2.txt\n",
      "22653 200\n",
      "2 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed2.txt\n",
      "22653 400\n",
      "2 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed2.txt\n",
      "22653 600\n",
      "2 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed2.txt\n",
      "22653 800\n",
      "2 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed2.txt\n",
      "22653 1000\n",
      "2 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed2.txt\n",
      "22653 3000\n",
      "2 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed2.txt\n",
      "22653 5000\n",
      "2 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed2.txt\n",
      "22653 10000\n",
      "2 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed2.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "3 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed3.txt\n",
      "26450 200\n",
      "3 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed3.txt\n",
      "26450 400\n",
      "3 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed3.txt\n",
      "26450 600\n",
      "3 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed3.txt\n",
      "26450 800\n",
      "3 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed3.txt\n",
      "26450 1000\n",
      "3 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed3.txt\n",
      "26450 3000\n",
      "3 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed3.txt\n",
      "26450 5000\n",
      "3 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed3.txt\n",
      "26450 10000\n",
      "3 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed3.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "3 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed3.txt\n",
      "22653 200\n",
      "3 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed3.txt\n",
      "22653 400\n",
      "3 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed3.txt\n",
      "22653 600\n",
      "3 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed3.txt\n",
      "22653 800\n",
      "3 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed3.txt\n",
      "22653 1000\n",
      "3 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed3.txt\n",
      "22653 3000\n",
      "3 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed3.txt\n",
      "22653 5000\n",
      "3 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed3.txt\n",
      "22653 10000\n",
      "3 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed3.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "4 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed4.txt\n",
      "26450 200\n",
      "4 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed4.txt\n",
      "26450 400\n",
      "4 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed4.txt\n",
      "26450 600\n",
      "4 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed4.txt\n",
      "26450 800\n",
      "4 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed4.txt\n",
      "26450 1000\n",
      "4 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed4.txt\n",
      "26450 3000\n",
      "4 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed4.txt\n",
      "26450 5000\n",
      "4 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed4.txt\n",
      "26450 10000\n",
      "4 RNA 15\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed4.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "4 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed4.txt\n",
      "22653 200\n",
      "4 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed4.txt\n",
      "22653 400\n",
      "4 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed4.txt\n",
      "22653 600\n",
      "4 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed4.txt\n",
      "22653 800\n",
      "4 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed4.txt\n",
      "22653 1000\n",
      "4 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed4.txt\n",
      "22653 3000\n",
      "4 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed4.txt\n",
      "22653 5000\n",
      "4 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed4.txt\n",
      "22653 10000\n",
      "4 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed4.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "5 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed5.txt\n",
      "26450 200\n",
      "5 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed5.txt\n",
      "26450 400\n",
      "5 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed5.txt\n",
      "26450 600\n",
      "5 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed5.txt\n",
      "26450 800\n",
      "5 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed5.txt\n",
      "26450 1000\n",
      "5 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed5.txt\n",
      "26450 3000\n",
      "5 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed5.txt\n",
      "26450 5000\n",
      "5 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed5.txt\n",
      "26450 10000\n",
      "5 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed5.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "5 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed5.txt\n",
      "22653 200\n",
      "5 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed5.txt\n",
      "22653 400\n",
      "5 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed5.txt\n",
      "22653 600\n",
      "5 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed5.txt\n",
      "22653 800\n",
      "5 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed5.txt\n",
      "22653 1000\n",
      "5 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed5.txt\n",
      "22653 3000\n",
      "5 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed5.txt\n",
      "22653 5000\n",
      "5 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed5.txt\n",
      "22653 10000\n",
      "5 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed5.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "6 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed6.txt\n",
      "26450 200\n",
      "6 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed6.txt\n",
      "26450 400\n",
      "6 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed6.txt\n",
      "26450 600\n",
      "6 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed6.txt\n",
      "26450 800\n",
      "6 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed6.txt\n",
      "26450 1000\n",
      "6 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed6.txt\n",
      "26450 3000\n",
      "6 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed6.txt\n",
      "26450 5000\n",
      "6 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed6.txt\n",
      "26450 10000\n",
      "6 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed6.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "6 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed6.txt\n",
      "22653 200\n",
      "6 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed6.txt\n",
      "22653 400\n",
      "6 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed6.txt\n",
      "22653 600\n",
      "6 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed6.txt\n",
      "22653 800\n",
      "6 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed6.txt\n",
      "22653 1000\n",
      "6 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed6.txt\n",
      "22653 3000\n",
      "6 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed6.txt\n",
      "22653 5000\n",
      "6 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed6.txt\n",
      "22653 10000\n",
      "6 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed6.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "7 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed7.txt\n",
      "26450 200\n",
      "7 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed7.txt\n",
      "26450 400\n",
      "7 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed7.txt\n",
      "26450 600\n",
      "7 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed7.txt\n",
      "26450 800\n",
      "7 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed7.txt\n",
      "26450 1000\n",
      "7 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed7.txt\n",
      "26450 3000\n",
      "7 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed7.txt\n",
      "26450 5000\n",
      "7 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed7.txt\n",
      "26450 10000\n",
      "7 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed7.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "7 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed7.txt\n",
      "22653 200\n",
      "7 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed7.txt\n",
      "22653 400\n",
      "7 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed7.txt\n",
      "22653 600\n",
      "7 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed7.txt\n",
      "22653 800\n",
      "7 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed7.txt\n",
      "22653 1000\n",
      "7 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed7.txt\n",
      "22653 3000\n",
      "7 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed7.txt\n",
      "22653 5000\n",
      "7 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed7.txt\n",
      "22653 10000\n",
      "7 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed7.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "8 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed8.txt\n",
      "26450 200\n",
      "8 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed8.txt\n",
      "26450 400\n",
      "8 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed8.txt\n",
      "26450 600\n",
      "8 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed8.txt\n",
      "26450 800\n",
      "8 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed8.txt\n",
      "26450 1000\n",
      "8 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed8.txt\n",
      "26450 3000\n",
      "8 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed8.txt\n",
      "26450 5000\n",
      "8 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed8.txt\n",
      "26450 10000\n",
      "8 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed8.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "8 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed8.txt\n",
      "22653 200\n",
      "8 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed8.txt\n",
      "22653 400\n",
      "8 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed8.txt\n",
      "22653 600\n",
      "8 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed8.txt\n",
      "22653 800\n",
      "8 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed8.txt\n",
      "22653 1000\n",
      "8 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed8.txt\n",
      "22653 3000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed8.txt\n",
      "22653 5000\n",
      "8 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed8.txt\n",
      "22653 10000\n",
      "8 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed8.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "9 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed9.txt\n",
      "26450 200\n",
      "9 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed9.txt\n",
      "26450 400\n",
      "9 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed9.txt\n",
      "26450 600\n",
      "9 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed9.txt\n",
      "26450 800\n",
      "9 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed9.txt\n",
      "26450 1000\n",
      "9 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed9.txt\n",
      "26450 3000\n",
      "9 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed9.txt\n",
      "26450 5000\n",
      "9 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed9.txt\n",
      "26450 10000\n",
      "9 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed9.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "9 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed9.txt\n",
      "22653 200\n",
      "9 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed9.txt\n",
      "22653 400\n",
      "9 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed9.txt\n",
      "22653 600\n",
      "9 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed9.txt\n",
      "22653 800\n",
      "9 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed9.txt\n",
      "22653 1000\n",
      "9 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed9.txt\n",
      "22653 3000\n",
      "9 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed9.txt\n",
      "22653 5000\n",
      "9 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed9.txt\n",
      "22653 10000\n",
      "9 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed9.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "10 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed10.txt\n",
      "26450 200\n",
      "10 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed10.txt\n",
      "26450 400\n",
      "10 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed10.txt\n",
      "26450 600\n",
      "10 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed10.txt\n",
      "26450 800\n",
      "10 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed10.txt\n",
      "26450 1000\n",
      "10 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed10.txt\n",
      "26450 3000\n",
      "10 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed10.txt\n",
      "26450 5000\n",
      "10 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed10.txt\n",
      "26450 10000\n",
      "10 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed10.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "10 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed10.txt\n",
      "22653 200\n",
      "10 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed10.txt\n",
      "22653 400\n",
      "10 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed10.txt\n",
      "22653 600\n",
      "10 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed10.txt\n",
      "22653 800\n",
      "10 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed10.txt\n",
      "22653 1000\n",
      "10 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed10.txt\n",
      "22653 3000\n",
      "10 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed10.txt\n",
      "22653 5000\n",
      "10 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed10.txt\n",
      "22653 10000\n",
      "10 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed10.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "11 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed11.txt\n",
      "26450 200\n",
      "11 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed11.txt\n",
      "26450 400\n",
      "11 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed11.txt\n",
      "26450 600\n",
      "11 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed11.txt\n",
      "26450 800\n",
      "11 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed11.txt\n",
      "26450 1000\n",
      "11 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed11.txt\n",
      "26450 3000\n",
      "11 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed11.txt\n",
      "26450 5000\n",
      "11 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed11.txt\n",
      "26450 10000\n",
      "11 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed11.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "11 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed11.txt\n",
      "22653 200\n",
      "11 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed11.txt\n",
      "22653 400\n",
      "11 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed11.txt\n",
      "22653 600\n",
      "11 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed11.txt\n",
      "22653 800\n",
      "11 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed11.txt\n",
      "22653 1000\n",
      "11 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed11.txt\n",
      "22653 3000\n",
      "11 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed11.txt\n",
      "22653 5000\n",
      "11 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed11.txt\n",
      "22653 10000\n",
      "11 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed11.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "12 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed12.txt\n",
      "26450 200\n",
      "12 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed12.txt\n",
      "26450 400\n",
      "12 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed12.txt\n",
      "26450 600\n",
      "12 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed12.txt\n",
      "26450 800\n",
      "12 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed12.txt\n",
      "26450 1000\n",
      "12 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed12.txt\n",
      "26450 3000\n",
      "12 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed12.txt\n",
      "26450 5000\n",
      "12 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed12.txt\n",
      "26450 10000\n",
      "12 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed12.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "12 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed12.txt\n",
      "22653 200\n",
      "12 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed12.txt\n",
      "22653 400\n",
      "12 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed12.txt\n",
      "22653 600\n",
      "12 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed12.txt\n",
      "22653 800\n",
      "12 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed12.txt\n",
      "22653 1000\n",
      "12 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed12.txt\n",
      "22653 3000\n",
      "12 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed12.txt\n",
      "22653 5000\n",
      "12 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed12.txt\n",
      "22653 10000\n",
      "12 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed12.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "13 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed13.txt\n",
      "26450 200\n",
      "13 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed13.txt\n",
      "26450 400\n",
      "13 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed13.txt\n",
      "26450 600\n",
      "13 RNA 0.8\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed13.txt\n",
      "26450 800\n",
      "13 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed13.txt\n",
      "26450 1000\n",
      "13 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed13.txt\n",
      "26450 3000\n",
      "13 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed13.txt\n",
      "26450 5000\n",
      "13 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed13.txt\n",
      "26450 10000\n",
      "13 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed13.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "13 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed13.txt\n",
      "22653 200\n",
      "13 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed13.txt\n",
      "22653 400\n",
      "13 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed13.txt\n",
      "22653 600\n",
      "13 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed13.txt\n",
      "22653 800\n",
      "13 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed13.txt\n",
      "22653 1000\n",
      "13 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed13.txt\n",
      "22653 3000\n",
      "13 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed13.txt\n",
      "22653 5000\n",
      "13 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed13.txt\n",
      "22653 10000\n",
      "13 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed13.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "14 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed14.txt\n",
      "26450 200\n",
      "14 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed14.txt\n",
      "26450 400\n",
      "14 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed14.txt\n",
      "26450 600\n",
      "14 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed14.txt\n",
      "26450 800\n",
      "14 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed14.txt\n",
      "26450 1000\n",
      "14 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed14.txt\n",
      "26450 3000\n",
      "14 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed14.txt\n",
      "26450 5000\n",
      "14 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed14.txt\n",
      "26450 10000\n",
      "14 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed14.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "14 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed14.txt\n",
      "22653 200\n",
      "14 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed14.txt\n",
      "22653 400\n",
      "14 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed14.txt\n",
      "22653 600\n",
      "14 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed14.txt\n",
      "22653 800\n",
      "14 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed14.txt\n",
      "22653 1000\n",
      "14 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed14.txt\n",
      "22653 3000\n",
      "14 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed14.txt\n",
      "22653 5000\n",
      "14 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed14.txt\n",
      "22653 10000\n",
      "14 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed14.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "15 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed15.txt\n",
      "26450 200\n",
      "15 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed15.txt\n",
      "26450 400\n",
      "15 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed15.txt\n",
      "26450 600\n",
      "15 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed15.txt\n",
      "26450 800\n",
      "15 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed15.txt\n",
      "26450 1000\n",
      "15 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed15.txt\n",
      "26450 3000\n",
      "15 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed15.txt\n",
      "26450 5000\n",
      "15 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed15.txt\n",
      "26450 10000\n",
      "15 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed15.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "15 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed15.txt\n",
      "22653 200\n",
      "15 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed15.txt\n",
      "22653 400\n",
      "15 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed15.txt\n",
      "22653 600\n",
      "15 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed15.txt\n",
      "22653 800\n",
      "15 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed15.txt\n",
      "22653 1000\n",
      "15 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed15.txt\n",
      "22653 3000\n",
      "15 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed15.txt\n",
      "22653 5000\n",
      "15 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed15.txt\n",
      "22653 10000\n",
      "15 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed15.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "16 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed16.txt\n",
      "26450 200\n",
      "16 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed16.txt\n",
      "26450 400\n",
      "16 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed16.txt\n",
      "26450 600\n",
      "16 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed16.txt\n",
      "26450 800\n",
      "16 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed16.txt\n",
      "26450 1000\n",
      "16 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed16.txt\n",
      "26450 3000\n",
      "16 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed16.txt\n",
      "26450 5000\n",
      "16 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed16.txt\n",
      "26450 10000\n",
      "16 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed16.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "16 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed16.txt\n",
      "22653 200\n",
      "16 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed16.txt\n",
      "22653 400\n",
      "16 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed16.txt\n",
      "22653 600\n",
      "16 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed16.txt\n",
      "22653 800\n",
      "16 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed16.txt\n",
      "22653 1000\n",
      "16 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed16.txt\n",
      "22653 3000\n",
      "16 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed16.txt\n",
      "22653 5000\n",
      "16 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed16.txt\n",
      "22653 10000\n",
      "16 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed16.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "17 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed17.txt\n",
      "26450 200\n",
      "17 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed17.txt\n",
      "26450 400\n",
      "17 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed17.txt\n",
      "26450 600\n",
      "17 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed17.txt\n",
      "26450 800\n",
      "17 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed17.txt\n",
      "26450 1000\n",
      "17 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed17.txt\n",
      "26450 3000\n",
      "17 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed17.txt\n",
      "26450 5000\n",
      "17 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed17.txt\n",
      "26450 10000\n",
      "17 RNA 15\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed17.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "17 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed17.txt\n",
      "22653 200\n",
      "17 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed17.txt\n",
      "22653 400\n",
      "17 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed17.txt\n",
      "22653 600\n",
      "17 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed17.txt\n",
      "22653 800\n",
      "17 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed17.txt\n",
      "22653 1000\n",
      "17 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed17.txt\n",
      "22653 3000\n",
      "17 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed17.txt\n",
      "22653 5000\n",
      "17 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed17.txt\n",
      "22653 10000\n",
      "17 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed17.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "18 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed18.txt\n",
      "26450 200\n",
      "18 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed18.txt\n",
      "26450 400\n",
      "18 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed18.txt\n",
      "26450 600\n",
      "18 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed18.txt\n",
      "26450 800\n",
      "18 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed18.txt\n",
      "26450 1000\n",
      "18 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed18.txt\n",
      "26450 3000\n",
      "18 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed18.txt\n",
      "26450 5000\n",
      "18 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed18.txt\n",
      "26450 10000\n",
      "18 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed18.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "18 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed18.txt\n",
      "22653 200\n",
      "18 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed18.txt\n",
      "22653 400\n",
      "18 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed18.txt\n",
      "22653 600\n",
      "18 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed18.txt\n",
      "22653 800\n",
      "18 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed18.txt\n",
      "22653 1000\n",
      "18 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed18.txt\n",
      "22653 3000\n",
      "18 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed18.txt\n",
      "22653 5000\n",
      "18 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed18.txt\n",
      "22653 10000\n",
      "18 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed18.txt\n",
      "22653 15000\n",
      "\n",
      "\n",
      "19 RNA 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.2k_seed19.txt\n",
      "26450 200\n",
      "19 RNA 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.4k_seed19.txt\n",
      "26450 400\n",
      "19 RNA 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.6k_seed19.txt\n",
      "26450 600\n",
      "19 RNA 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_0.8k_seed19.txt\n",
      "26450 800\n",
      "19 RNA 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_1k_seed19.txt\n",
      "26450 1000\n",
      "19 RNA 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_3k_seed19.txt\n",
      "26450 3000\n",
      "19 RNA 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_5k_seed19.txt\n",
      "26450 5000\n",
      "19 RNA 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_10k_seed19.txt\n",
      "26450 10000\n",
      "19 RNA 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/RNA/cells_15k_seed19.txt\n",
      "26450 15000\n",
      "\n",
      "\n",
      "19 ATAC 0.2\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.2k_seed19.txt\n",
      "22653 200\n",
      "19 ATAC 0.4\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.4k_seed19.txt\n",
      "22653 400\n",
      "19 ATAC 0.6\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.6k_seed19.txt\n",
      "22653 600\n",
      "19 ATAC 0.8\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_0.8k_seed19.txt\n",
      "22653 800\n",
      "19 ATAC 1\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_1k_seed19.txt\n",
      "22653 1000\n",
      "19 ATAC 3\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_3k_seed19.txt\n",
      "22653 3000\n",
      "19 ATAC 5\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_5k_seed19.txt\n",
      "22653 5000\n",
      "19 ATAC 10\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_10k_seed19.txt\n",
      "22653 10000\n",
      "19 ATAC 15\n",
      "/gpfs/gibbs/pi/zhao/xs272/Multiomics/sc_benchmark/data/BMMC/ATAC/cells_15k_seed19.txt\n",
      "22653 15000\n"
     ]
    }
   ],
   "source": [
    "# paths\n",
    "data_path = '/gpfs/gibbs/pi/zhao/yw599/Multiome/data_pp/BMMC'\n",
    "folders = ['RNA', 'ATAC']\n",
    "ks = list(np.round(np.arange(0.2, 1., step=0.2), 1)) + [1,3,5,10,15]\n",
    "for seed in range(20):\n",
    "    random.seed(seed)\n",
    "    np.random.seed(seed)\n",
    "    for folder in folders:\n",
    "        print('\\n')\n",
    "        for k in ks:\n",
    "            print(seed, folder, k)\n",
    "            cells = np.loadtxt(path.join(data_path, folder, 'cells.txt'), dtype=object, delimiter='SPACEISPARTOFTHENAME')\n",
    "            cells_tmp = random.sample(list(cells), int(1000*k))\n",
    "            savepath = path.join(data_path, folder, 'cells_%dk_seed%s.txt' % (k, seed) if k>=1 else 'cells_%.1fk_seed%s.txt' % (k, seed))\n",
    "            print(savepath)\n",
    "            print(len(cells), len(cells_tmp))\n",
    "            np.savetxt(savepath, cells_tmp, fmt='%s')\n",
    "            if folder == 'ATAC':\n",
    "                savepath = path.join(data_path, 'ATAC.gene.activities', 'cells_%dk_seed%s.txt' % (k, seed) if k>=1 else 'cells_%.1fk_seed%s.txt' % (k, seed))\n",
    "                np.savetxt(savepath, cells_tmp, fmt='%s')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d1012b1",
   "metadata": {},
   "source": [
    "## Subsample number of multiome RNA and ATAC cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "23e1c922",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 0.2\n",
      "18467 200\n",
      "0 0.4\n",
      "18467 400\n",
      "0 0.6\n",
      "18467 600\n",
      "0 0.8\n",
      "18467 800\n",
      "0 1\n",
      "18467 1000\n",
      "0 3\n",
      "18467 3000\n",
      "0 5\n",
      "18467 5000\n",
      "0 10\n",
      "18467 10000\n",
      "0 15\n",
      "18467 15000\n",
      "1 0.2\n",
      "18467 200\n",
      "1 0.4\n",
      "18467 400\n",
      "1 0.6\n",
      "18467 600\n",
      "1 0.8\n",
      "18467 800\n",
      "1 1\n",
      "18467 1000\n",
      "1 3\n",
      "18467 3000\n",
      "1 5\n",
      "18467 5000\n",
      "1 10\n",
      "18467 10000\n",
      "1 15\n",
      "18467 15000\n",
      "2 0.2\n",
      "18467 200\n",
      "2 0.4\n",
      "18467 400\n",
      "2 0.6\n",
      "18467 600\n",
      "2 0.8\n",
      "18467 800\n",
      "2 1\n",
      "18467 1000\n",
      "2 3\n",
      "18467 3000\n",
      "2 5\n",
      "18467 5000\n",
      "2 10\n",
      "18467 10000\n",
      "2 15\n",
      "18467 15000\n",
      "3 0.2\n",
      "18467 200\n",
      "3 0.4\n",
      "18467 400\n",
      "3 0.6\n",
      "18467 600\n",
      "3 0.8\n",
      "18467 800\n",
      "3 1\n",
      "18467 1000\n",
      "3 3\n",
      "18467 3000\n",
      "3 5\n",
      "18467 5000\n",
      "3 10\n",
      "18467 10000\n",
      "3 15\n",
      "18467 15000\n",
      "4 0.2\n",
      "18467 200\n",
      "4 0.4\n",
      "18467 400\n",
      "4 0.6\n",
      "18467 600\n",
      "4 0.8\n",
      "18467 800\n",
      "4 1\n",
      "18467 1000\n",
      "4 3\n",
      "18467 3000\n",
      "4 5\n",
      "18467 5000\n",
      "4 10\n",
      "18467 10000\n",
      "4 15\n",
      "18467 15000\n",
      "5 0.2\n",
      "18467 200\n",
      "5 0.4\n",
      "18467 400\n",
      "5 0.6\n",
      "18467 600\n",
      "5 0.8\n",
      "18467 800\n",
      "5 1\n",
      "18467 1000\n",
      "5 3\n",
      "18467 3000\n",
      "5 5\n",
      "18467 5000\n",
      "5 10\n",
      "18467 10000\n",
      "5 15\n",
      "18467 15000\n",
      "6 0.2\n",
      "18467 200\n",
      "6 0.4\n",
      "18467 400\n",
      "6 0.6\n",
      "18467 600\n",
      "6 0.8\n",
      "18467 800\n",
      "6 1\n",
      "18467 1000\n",
      "6 3\n",
      "18467 3000\n",
      "6 5\n",
      "18467 5000\n",
      "6 10\n",
      "18467 10000\n",
      "6 15\n",
      "18467 15000\n",
      "7 0.2\n",
      "18467 200\n",
      "7 0.4\n",
      "18467 400\n",
      "7 0.6\n",
      "18467 600\n",
      "7 0.8\n",
      "18467 800\n",
      "7 1\n",
      "18467 1000\n",
      "7 3\n",
      "18467 3000\n",
      "7 5\n",
      "18467 5000\n",
      "7 10\n",
      "18467 10000\n",
      "7 15\n",
      "18467 15000\n",
      "8 0.2\n",
      "18467 200\n",
      "8 0.4\n",
      "18467 400\n",
      "8 0.6\n",
      "18467 600\n",
      "8 0.8\n",
      "18467 800\n",
      "8 1\n",
      "18467 1000\n",
      "8 3\n",
      "18467 3000\n",
      "8 5\n",
      "18467 5000\n",
      "8 10\n",
      "18467 10000\n",
      "8 15\n",
      "18467 15000\n",
      "9 0.2\n",
      "18467 200\n",
      "9 0.4\n",
      "18467 400\n",
      "9 0.6\n",
      "18467 600\n",
      "9 0.8\n",
      "18467 800\n",
      "9 1\n",
      "18467 1000\n",
      "9 3\n",
      "18467 3000\n",
      "9 5\n",
      "18467 5000\n",
      "9 10\n",
      "18467 10000\n",
      "9 15\n",
      "18467 15000\n",
      "10 0.2\n",
      "18467 200\n",
      "10 0.4\n",
      "18467 400\n",
      "10 0.6\n",
      "18467 600\n",
      "10 0.8\n",
      "18467 800\n",
      "10 1\n",
      "18467 1000\n",
      "10 3\n",
      "18467 3000\n",
      "10 5\n",
      "18467 5000\n",
      "10 10\n",
      "18467 10000\n",
      "10 15\n",
      "18467 15000\n",
      "11 0.2\n",
      "18467 200\n",
      "11 0.4\n",
      "18467 400\n",
      "11 0.6\n",
      "18467 600\n",
      "11 0.8\n",
      "18467 800\n",
      "11 1\n",
      "18467 1000\n",
      "11 3\n",
      "18467 3000\n",
      "11 5\n",
      "18467 5000\n",
      "11 10\n",
      "18467 10000\n",
      "11 15\n",
      "18467 15000\n",
      "12 0.2\n",
      "18467 200\n",
      "12 0.4\n",
      "18467 400\n",
      "12 0.6\n",
      "18467 600\n",
      "12 0.8\n",
      "18467 800\n",
      "12 1\n",
      "18467 1000\n",
      "12 3\n",
      "18467 3000\n",
      "12 5\n",
      "18467 5000\n",
      "12 10\n",
      "18467 10000\n",
      "12 15\n",
      "18467 15000\n",
      "13 0.2\n",
      "18467 200\n",
      "13 0.4\n",
      "18467 400\n",
      "13 0.6\n",
      "18467 600\n",
      "13 0.8\n",
      "18467 800\n",
      "13 1\n",
      "18467 1000\n",
      "13 3\n",
      "18467 3000\n",
      "13 5\n",
      "18467 5000\n",
      "13 10\n",
      "18467 10000\n",
      "13 15\n",
      "18467 15000\n",
      "14 0.2\n",
      "18467 200\n",
      "14 0.4\n",
      "18467 400\n",
      "14 0.6\n",
      "18467 600\n",
      "14 0.8\n",
      "18467 800\n",
      "14 1\n",
      "18467 1000\n",
      "14 3\n",
      "18467 3000\n",
      "14 5\n",
      "18467 5000\n",
      "14 10\n",
      "18467 10000\n",
      "14 15\n",
      "18467 15000\n",
      "15 0.2\n",
      "18467 200\n",
      "15 0.4\n",
      "18467 400\n",
      "15 0.6\n",
      "18467 600\n",
      "15 0.8\n",
      "18467 800\n",
      "15 1\n",
      "18467 1000\n",
      "15 3\n",
      "18467 3000\n",
      "15 5\n",
      "18467 5000\n",
      "15 10\n",
      "18467 10000\n",
      "15 15\n",
      "18467 15000\n",
      "16 0.2\n",
      "18467 200\n",
      "16 0.4\n",
      "18467 400\n",
      "16 0.6\n",
      "18467 600\n",
      "16 0.8\n",
      "18467 800\n",
      "16 1\n",
      "18467 1000\n",
      "16 3\n",
      "18467 3000\n",
      "16 5\n",
      "18467 5000\n",
      "16 10\n",
      "18467 10000\n",
      "16 15\n",
      "18467 15000\n",
      "17 0.2\n",
      "18467 200\n",
      "17 0.4\n",
      "18467 400\n",
      "17 0.6\n",
      "18467 600\n",
      "17 0.8\n",
      "18467 800\n",
      "17 1\n",
      "18467 1000\n",
      "17 3\n",
      "18467 3000\n",
      "17 5\n",
      "18467 5000\n",
      "17 10\n",
      "18467 10000\n",
      "17 15\n",
      "18467 15000\n",
      "18 0.2\n",
      "18467 200\n",
      "18 0.4\n",
      "18467 400\n",
      "18 0.6\n",
      "18467 600\n",
      "18 0.8\n",
      "18467 800\n",
      "18 1\n",
      "18467 1000\n",
      "18 3\n",
      "18467 3000\n",
      "18 5\n",
      "18467 5000\n",
      "18 10\n",
      "18467 10000\n",
      "18 15\n",
      "18467 15000\n",
      "19 0.2\n",
      "18467 200\n",
      "19 0.4\n",
      "18467 400\n",
      "19 0.6\n",
      "18467 600\n",
      "19 0.8\n",
      "18467 800\n",
      "19 1\n",
      "18467 1000\n",
      "19 3\n",
      "18467 3000\n",
      "19 5\n",
      "18467 5000\n",
      "19 10\n",
      "18467 10000\n",
      "19 15\n",
      "18467 15000\n"
     ]
    }
   ],
   "source": [
    "# paths\n",
    "data_path = '/gpfs/gibbs/pi/zhao/yw599/Multiome/data_pp/BMMC/multiome'\n",
    "ks = list(np.round(np.arange(0.2, 1., step=0.2), 1)) + [1,3,5,10,15]\n",
    "for seed in range(20):\n",
    "    random.seed(seed)\n",
    "    np.random.seed(seed)\n",
    "    for k in ks:\n",
    "        print(seed, k)\n",
    "        cells = np.loadtxt(path.join(data_path, 'RNA', 'cells.txt'), dtype=object, delimiter='SPACEISPARTOFTHENAME')\n",
    "        cells_tmp = random.sample(list(cells), int(1000*k))\n",
    "        print(len(cells), len(cells_tmp))\n",
    "        np.savetxt(path.join(data_path, 'RNA', 'cells_%dk_seed%s.txt' % (k, seed) if k>=1 else 'cells_%.1fk_seed%s.txt' % (k, seed)), cells_tmp, fmt='%s')\n",
    "        np.savetxt(path.join(data_path, 'ATAC', 'cells_%dk_seed%s.txt' % (k, seed) if k>=1 else 'cells_%.1fk_seed%s.txt' % (k, seed)), cells_tmp, fmt='%s')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac79783",
   "metadata": {},
   "source": [
    "## Generate mislabelled RNA annotation txt files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0a079dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/gpfs/gibbs/pi/zhao/yw599/Multiome/data_pp/BMMC/RNA'\n",
    "anno = np.loadtxt(path.join(data_path, 'annotations.txt'), dtype=object, delimiter='SPACEISPARTOFTHENAME')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "632543ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "dic = {}\n",
    "for ct in set(anno):\n",
    "    dic[ct] = np.sum(anno == ct)/len(anno)\n",
    "prob = pd.DataFrame.from_dict(dic, orient='index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e28f3d72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mislabel(anno, perc):\n",
    "    idx = np.random.choice(len(anno), round(len(anno)*perc), replace=False)\n",
    "    tmp = anno.copy()\n",
    "    for i in idx:\n",
    "        ct = anno[i]\n",
    "        prob_ = prob[prob.index != ct]\n",
    "        ct_ = prob_.index.tolist()\n",
    "        prob_ = prob_.values.reshape(-1)\n",
    "        tmp[i] = random.choices(ct_, prob_, k=1)[0]\n",
    "    return tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7fd9ea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # paths\n",
    "# ks = [10,20,30,40,50,60,70,80,90,100]\n",
    "# for seed in range(20):\n",
    "#     random.seed(seed)\n",
    "#     np.random.seed(seed)\n",
    "#     for k in ks:\n",
    "#         print(seed, k)\n",
    "#         anno_tmp = mislabel(anno, k/100.)\n",
    "#         print(np.sum(anno != anno_tmp) / len(anno))\n",
    "#         np.savetxt(path.join(data_path, 'annotations_mis_percent_%d_seed%s.txt' % (k, seed)), anno_tmp, fmt='%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "74845b40",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "0.009981096408317581\n",
      "5\n",
      "0.049981096408317584\n",
      "10\n",
      "0.1\n",
      "20\n",
      "0.2\n",
      "30\n",
      "0.3\n",
      "40\n",
      "0.4\n",
      "50\n",
      "0.5\n",
      "60\n",
      "0.6\n",
      "70\n",
      "0.7\n",
      "80\n",
      "0.8\n",
      "90\n",
      "0.9\n",
      "100\n",
      "1.0\n"
     ]
    }
   ],
   "source": [
    "# paths\n",
    "random.seed(10)\n",
    "np.random.seed(10)\n",
    "ks = [10,20,30,40,50,60,70,80,90,100]\n",
    "for k in ks:\n",
    "    print(k)\n",
    "    anno_tmp = mislabel(anno, k/100.)\n",
    "    print(np.sum(anno != anno_tmp) / len(anno))\n",
    "    np.savetxt(path.join(data_path, 'annotations_mis_percent_%d.txt' % k), anno_tmp, fmt='%s')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90eb3be9",
   "metadata": {},
   "source": [
    "## Generate cells.txt with partial cell tyes for RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8c8415da",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/gpfs/gibbs/pi/zhao/yw599/Multiome/data_pp/BMMC'\n",
    "cells = np.loadtxt(path.join(data_path, 'RNA', 'cells.txt'), dtype=object, delimiter='SPACEISPARTOFTHENAME')\n",
    "anno = np.loadtxt(path.join(data_path, 'RNA', 'annotations.txt'), dtype=object, delimiter='SPACEISPARTOFTHENAME')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "da55e371",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(anno, index=cells, columns=['anno'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9d6cb65f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26450, 1)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "64f6d1e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CD8+ T                 4267\n",
       "CD14+ Mono             3404\n",
       "NK                     3158\n",
       "CD4+ T naive           2368\n",
       "Naive CD20+ B          1965\n",
       "CD4+ T activated       1685\n",
       "Erythroblast           1237\n",
       "Transitional B         1138\n",
       "CD8+ T naive           1012\n",
       "Lymph prog              682\n",
       "CD16+ Mono              653\n",
       "Proerythroblast         646\n",
       "B1 B                    642\n",
       "pDC                     632\n",
       "ILC                     620\n",
       "G/M prog                577\n",
       "HSC                     542\n",
       "cDC2                    384\n",
       "Normoblast              375\n",
       "MK/E prog               371\n",
       "Plasma cell              51\n",
       "ID2-hi myeloid prog      41\n",
       "Name: anno, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.anno.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "da111696",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = df.anno.value_counts()\n",
    "candidate_ct = list(tmp[(tmp < 1000) & (tmp > 200)].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "67b9a79d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Lymph prog',\n",
       " 'CD16+ Mono',\n",
       " 'Proerythroblast',\n",
       " 'B1 B',\n",
       " 'pDC',\n",
       " 'ILC',\n",
       " 'G/M prog',\n",
       " 'HSC',\n",
       " 'cDC2',\n",
       " 'Normoblast',\n",
       " 'MK/E prog']"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a4eefc9f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 2\n",
      "0 4\n",
      "0 6\n",
      "0 8\n",
      "0 10\n",
      "1 2\n",
      "1 4\n",
      "1 6\n",
      "1 8\n",
      "1 10\n",
      "2 2\n",
      "2 4\n",
      "2 6\n",
      "2 8\n",
      "2 10\n",
      "3 2\n",
      "3 4\n",
      "3 6\n",
      "3 8\n",
      "3 10\n",
      "4 2\n",
      "4 4\n",
      "4 6\n",
      "4 8\n",
      "4 10\n",
      "5 2\n",
      "5 4\n",
      "5 6\n",
      "5 8\n",
      "5 10\n",
      "6 2\n",
      "6 4\n",
      "6 6\n",
      "6 8\n",
      "6 10\n",
      "7 2\n",
      "7 4\n",
      "7 6\n",
      "7 8\n",
      "7 10\n",
      "8 2\n",
      "8 4\n",
      "8 6\n",
      "8 8\n",
      "8 10\n",
      "9 2\n",
      "9 4\n",
      "9 6\n",
      "9 8\n",
      "9 10\n",
      "10 2\n",
      "10 4\n",
      "10 6\n",
      "10 8\n",
      "10 10\n",
      "11 2\n",
      "11 4\n",
      "11 6\n",
      "11 8\n",
      "11 10\n",
      "12 2\n",
      "12 4\n",
      "12 6\n",
      "12 8\n",
      "12 10\n",
      "13 2\n",
      "13 4\n",
      "13 6\n",
      "13 8\n",
      "13 10\n",
      "14 2\n",
      "14 4\n",
      "14 6\n",
      "14 8\n",
      "14 10\n",
      "15 2\n",
      "15 4\n",
      "15 6\n",
      "15 8\n",
      "15 10\n",
      "16 2\n",
      "16 4\n",
      "16 6\n",
      "16 8\n",
      "16 10\n",
      "17 2\n",
      "17 4\n",
      "17 6\n",
      "17 8\n",
      "17 10\n",
      "18 2\n",
      "18 4\n",
      "18 6\n",
      "18 8\n",
      "18 10\n",
      "19 2\n",
      "19 4\n",
      "19 6\n",
      "19 8\n",
      "19 10\n"
     ]
    }
   ],
   "source": [
    "# paths\n",
    "ks = [2,4,6,8,10]\n",
    "for seed in range(20):\n",
    "    random.seed(seed)\n",
    "    np.random.seed(seed)\n",
    "    cells_10k = np.loadtxt(path.join(data_path, 'RNA', 'cells_10k_seed%s.txt' % seed), dtype=object, delimiter='SPACEISPARTOFTHENAME')\n",
    "    for k in ks:\n",
    "        print(seed, k)\n",
    "        selected_ct = random.sample(candidate_ct, k)\n",
    "        df_10k = df.loc[cells_10k]\n",
    "        cells_tmp = list(df_10k[~df_10k.anno.isin(selected_ct)].index)\n",
    "        savepath = path.join(data_path, 'RNA', 'cells_10k_%srm_seed%s.txt' % (k, seed))\n",
    "        np.savetxt(savepath, cells_tmp, fmt='%s')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59b1db5f",
   "metadata": {},
   "source": [
    "## Generate cells.txt with partial cell tyes for multiome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "32e0e11c",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/gpfs/gibbs/pi/zhao/yw599/Multiome/data_pp/BMMC/multiome'\n",
    "cells = np.loadtxt(path.join(data_path, 'RNA', 'cells.txt'), dtype=object, delimiter='SPACEISPARTOFTHENAME')\n",
    "anno = np.loadtxt(path.join(data_path, 'RNA', 'annotations.txt'), dtype=object, delimiter='SPACEISPARTOFTHENAME')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "cc7ca2d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(anno, index=cells, columns=['anno'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d0e94bac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18467, 1)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b681854e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CD14+ Mono             3243\n",
       "NK                     1984\n",
       "CD4+ T activated       1901\n",
       "Erythroblast           1749\n",
       "CD8+ T                 1653\n",
       "Naive CD20+ B          1301\n",
       "CD4+ T naive           1198\n",
       "Transitional B          809\n",
       "B1 B                    688\n",
       "Normoblast              630\n",
       "CD16+ Mono              526\n",
       "Proerythroblast         472\n",
       "Lymph prog              452\n",
       "G/M prog                341\n",
       "pDC                     338\n",
       "MK/E prog               298\n",
       "Plasma cell             286\n",
       "HSC                     230\n",
       "cDC2                    181\n",
       "ILC                     120\n",
       "ID2-hi myeloid prog      67\n",
       "Name: anno, dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.anno.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e33b81da",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp = df.anno.value_counts()\n",
    "candidate_ct = list(tmp[(tmp < 800) & (tmp > 100)].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2013bf9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['B1 B',\n",
       " 'Normoblast',\n",
       " 'CD16+ Mono',\n",
       " 'Proerythroblast',\n",
       " 'Lymph prog',\n",
       " 'G/M prog',\n",
       " 'pDC',\n",
       " 'MK/E prog',\n",
       " 'Plasma cell',\n",
       " 'HSC',\n",
       " 'cDC2',\n",
       " 'ILC']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "candidate_ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "33d7ecb0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 2\n",
      "0 4\n",
      "0 6\n",
      "0 8\n",
      "0 10\n",
      "1 2\n",
      "1 4\n",
      "1 6\n",
      "1 8\n",
      "1 10\n",
      "2 2\n",
      "2 4\n",
      "2 6\n",
      "2 8\n",
      "2 10\n",
      "3 2\n",
      "3 4\n",
      "3 6\n",
      "3 8\n",
      "3 10\n",
      "4 2\n",
      "4 4\n",
      "4 6\n",
      "4 8\n",
      "4 10\n",
      "5 2\n",
      "5 4\n",
      "5 6\n",
      "5 8\n",
      "5 10\n",
      "6 2\n",
      "6 4\n",
      "6 6\n",
      "6 8\n",
      "6 10\n",
      "7 2\n",
      "7 4\n",
      "7 6\n",
      "7 8\n",
      "7 10\n",
      "8 2\n",
      "8 4\n",
      "8 6\n",
      "8 8\n",
      "8 10\n",
      "9 2\n",
      "9 4\n",
      "9 6\n",
      "9 8\n",
      "9 10\n",
      "10 2\n",
      "10 4\n",
      "10 6\n",
      "10 8\n",
      "10 10\n",
      "11 2\n",
      "11 4\n",
      "11 6\n",
      "11 8\n",
      "11 10\n",
      "12 2\n",
      "12 4\n",
      "12 6\n",
      "12 8\n",
      "12 10\n",
      "13 2\n",
      "13 4\n",
      "13 6\n",
      "13 8\n",
      "13 10\n",
      "14 2\n",
      "14 4\n",
      "14 6\n",
      "14 8\n",
      "14 10\n",
      "15 2\n",
      "15 4\n",
      "15 6\n",
      "15 8\n",
      "15 10\n",
      "16 2\n",
      "16 4\n",
      "16 6\n",
      "16 8\n",
      "16 10\n",
      "17 2\n",
      "17 4\n",
      "17 6\n",
      "17 8\n",
      "17 10\n",
      "18 2\n",
      "18 4\n",
      "18 6\n",
      "18 8\n",
      "18 10\n",
      "19 2\n",
      "19 4\n",
      "19 6\n",
      "19 8\n",
      "19 10\n"
     ]
    }
   ],
   "source": [
    "# paths\n",
    "ks = [2,4,6,8,10]\n",
    "for seed in range(20):\n",
    "    random.seed(seed)\n",
    "    np.random.seed(seed)\n",
    "    cells_10k = np.loadtxt(path.join(data_path, 'RNA', 'cells_10k.txt'), dtype=object, delimiter='SPACEISPARTOFTHENAME')\n",
    "    for k in ks:\n",
    "        print(seed, k)\n",
    "        selected_ct = random.sample(candidate_ct, k)\n",
    "        df_10k = df.loc[cells_10k]\n",
    "        cells_tmp = list(df_10k[~df_10k.anno.isin(selected_ct)].index)\n",
    "        savepath = path.join(data_path, 'RNA', 'cells_10k_%srm_seed%s.txt' % (k, seed))\n",
    "        np.savetxt(savepath, cells_tmp, fmt='%s')\n",
    "        savepath = path.join(data_path, 'ATAC', 'cells_10k_%srm_seed%s.txt' % (k, seed))\n",
    "        np.savetxt(savepath, cells_tmp, fmt='%s')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7301c9",
   "metadata": {},
   "source": [
    "## Generate downsampled data matrix (switch kernel to R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f8bfdaa8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Matrix)\n",
    "library(DropletUtils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fd99681",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/gpfs/gibbs/pi/zhao/yw599/Multiome/data_pp/BMMC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "28f5d98e",
   "metadata": {},
   "outputs": [],
   "source": [
    "depth <- c(0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9)\n",
    "data_type <- c('RNA', 'ATAC', 'ATAC.gene.activities', 'multiome/RNA', 'multiome/ATAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b9a27cee",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (d in data_type){\n",
    "    count <- readMM(file.path(data_path, d, \"counts.mtx\"))\n",
    "    for (p in depth){\n",
    "        tmp <- DropletUtils::downsampleMatrix(t(count), prop=p)\n",
    "        writeMM(t(tmp), file.path(data_path, d, paste0(\"counts_\", p, \".mtx\")))\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "374865b1",
   "metadata": {},
   "source": [
    "## Calculate unimodal ATAC gene activity (switch kernel to R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6fc6cdbd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: survcomp\n",
      "\n",
      "Loading required package: survival\n",
      "\n",
      "Loading required package: prodlim\n",
      "\n",
      "Loading required package: biomaRt\n",
      "\n",
      "Loading required package: iC10\n",
      "\n",
      "Loading required package: pamr\n",
      "\n",
      "Loading required package: cluster\n",
      "\n",
      "Warning message:\n",
      "“package ‘cluster’ was built under R version 4.2.1”\n",
      "Loading required package: impute\n",
      "\n",
      "Loading required package: iC10TrainingData\n",
      "\n",
      "Loading required package: AIMS\n",
      "\n",
      "Loading required package: e1071\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Warning message:\n",
      "“package ‘BiocGenerics’ was built under R version 4.2.1”\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "setwd('/gpfs/gibbs/pi/zhao/yw599/Multiome/data_pp/BMMC')\n",
    "library(Matrix)\n",
    "library(readr)\n",
    "source('/gpfs/gibbs/pi/zhao/yw599/Multiome/utils.r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1197c3ec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: ensembldb\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Warning message:\n",
      "“package ‘IRanges’ was built under R version 4.2.1”\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicFeatures\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationDbi’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "Loading required package: AnnotationFilter\n",
      "\n",
      "\n",
      "Attaching package: 'ensembldb'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "The following object is masked from 'package:stats':\n",
      "\n",
      "    filter\n",
      "\n",
      "\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n",
      "Warning message in .merge_two_Seqinfo_objects(x, y):\n",
      "\"The 2 combined objects have no sequence levels in common. (Use\n",
      "  suppressWarnings() to suppress this warning.)\"\n"
     ]
    }
   ],
   "source": [
    "# genome annotations\n",
    "library(EnsDb.Hsapiens.v86) # hg38\n",
    "library(Signac)\n",
    "annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)\n",
    "seqlevelsStyle(annotations) <- 'UCSC'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "01cf6107",
   "metadata": {},
   "outputs": [],
   "source": [
    "frag.path <- \"/gpfs/gibbs/pi/zhao/yw599/Multiome/data/BMMC/post_competition/multiome\"\n",
    "frag.files <- c(\n",
    "                s2d4=\"s2d4/atac_fragments.tsv.gz\",\n",
    "                s2d5=\"s2d5/atac_fragments.tsv.gz\",\n",
    "                s3d7=\"s3d7/atac_fragments.tsv.gz\",\n",
    "                s4d8=\"s4d8/atac_fragments.tsv.gz\",\n",
    "                s3d6=\"s3d6/atac_fragments.tsv.gz\"\n",
    "                )\n",
    "barcode_suffix <- c(\n",
    "                s2d4=\"s2d4\",\n",
    "                s2d5=\"6-s2d5\",\n",
    "                s3d7=\"12-s3d7\",\n",
    "                s4d8=\"14-s4d8\",\n",
    "                s3d6='8-s3d6')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "46fbc23d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "as(<dgTMatrix>, \"dgCMatrix\") is deprecated since Matrix 1.5-0; do as(., \"CsparseMatrix\") instead\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# read previously processed data\n",
    "single.atac.data <- load.data.folder('ATAC')\n",
    "single.atac.count <- single.atac.data$count\n",
    "single.atac.anno <- single.atac.data$annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "95684e03",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch <- sapply(colnames(single.atac.count), function(x) substr(x, (nchar(x) - 3), nchar(x)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0d5c1162",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"s2d4\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n",
      "Warning message:\n",
      "\"Keys should be one or more alphanumeric characters followed by an underscore, setting key from atac to atac_\"\n",
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 19607  6045\n",
      "[1] \"s2d5\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n",
      "Warning message:\n",
      "\"Keys should be one or more alphanumeric characters followed by an underscore, setting key from atac to atac_\"\n",
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 19607  4844\n",
      "[1] \"s3d7\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n",
      "Warning message:\n",
      "\"Keys should be one or more alphanumeric characters followed by an underscore, setting key from atac to atac_\"\n",
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 19607  1690\n",
      "[1] \"s4d8\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n",
      "Warning message:\n",
      "\"Keys should be one or more alphanumeric characters followed by an underscore, setting key from atac to atac_\"\n",
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 19607  9840\n",
      "[1] \"s3d6\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n",
      "Warning message:\n",
      "\"Keys should be one or more alphanumeric characters followed by an underscore, setting key from atac to atac_\"\n",
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 19607  1679\n"
     ]
    }
   ],
   "source": [
    "single.atac.ga.counts.list <- list()\n",
    "for (tissue in names(frag.files)) {\n",
    "    print(tissue)\n",
    "    selected.id <- which(batch == tissue)\n",
    "    selected.single.count <- single.atac.count[, selected.id]\n",
    "    colnames(selected.single.count) <- gsub(barcode_suffix[tissue], '1', colnames(selected.single.count))\n",
    "    selected.frag.path <- file.path(frag.path, frag.files[tissue])\n",
    "    selected.ga <- convert.gene.activtity(selected.single.count, selected.frag.path, annotations, genome='hg38')\n",
    "    colnames(selected.ga) <- gsub('1', barcode_suffix[tissue], colnames(selected.ga))\n",
    "    print(dim(selected.ga))\n",
    "    single.atac.ga.counts.list[[tissue]] <- selected.ga\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1ebc3b5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "single.atac.ga.counts <- do.call(cbind, single.atac.ga.counts.list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a9dcb0b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Length1, Length2, Length intersection, Identical:\"\n",
      "[1] \"24332 24332 24332 TRUE\"\n"
     ]
    }
   ],
   "source": [
    "# Check cell barcodes\n",
    "print.names.match(colnames(single.atac.ga.counts), colnames(single.atac.count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "50fd49ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Length1, Length2, Length intersection, Identical:\"\n",
      "[1] \"19607 13431 11141 FALSE\"\n"
     ]
    }
   ],
   "source": [
    "# Check features\n",
    "print.names.match(rownames(single.atac.ga.counts), readLines('RNA/genes.txt'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "eec15750",
   "metadata": {},
   "outputs": [],
   "source": [
    "path <- 'ATAC-gene-activities'\n",
    "write.counts.transpose(single.atac.ga.counts, path, anno=single.atac.anno)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d232fb2d",
   "metadata": {},
   "source": [
    "## Calculate unimodal ATAC gene activity (switch kernel to R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "10068dd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "frag.path <- \"/gpfs/gibbs/pi/zhao/yw599/Multiome/data/BMMC/post_competition/multiome\"\n",
    "frag.files <- c(\n",
    "                s1d1=\"s1d1/atac_fragments.tsv.gz\",\n",
    "                s2d1=\"s2d1/atac_fragments.tsv.gz\",\n",
    "                s4d1=\"s4d1/atac_fragments.tsv.gz\"\n",
    "                )\n",
    "barcode_suffix <- c(\n",
    "                s1d1=\"1-s1d1\",\n",
    "                s2d1=\"4-s2d1\",\n",
    "                s4d1=\"13-s4d1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "aa452e0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read previously processed data\n",
    "multi.atac.data <- load.data.folder('multiome/ATAC')\n",
    "multi.atac.count <- multi.atac.data$counts\n",
    "multi.atac.anno <- multi.atac.data$annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "171ad5eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "multi.atac.count <- multi.atac.data$count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5546bb06",
   "metadata": {},
   "outputs": [],
   "source": [
    "batch <- sapply(colnames(multi.atac.count), function(x) substr(x, (nchar(x) - 3), nchar(x)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1a8f6db5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "batch\n",
       "s1d1 s2d1 s4d1 \n",
       "6224 4220 8023 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table(batch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "818e2ec1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"s1d1\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n",
      "Warning message:\n",
      "\"Keys should be one or more alphanumeric characters followed by an underscore, setting key from atac to atac_\"\n",
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 19607  6221\n",
      "[1] \"s2d1\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n",
      "Warning message:\n",
      "\"Keys should be one or more alphanumeric characters followed by an underscore, setting key from atac to atac_\"\n",
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 19607  4180\n",
      "[1] \"s4d1\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing hash\n",
      "\n",
      "Warning message:\n",
      "\"Keys should be one or more alphanumeric characters followed by an underscore, setting key from atac to atac_\"\n",
      "Extracting gene coordinates\n",
      "\n",
      "Extracting reads overlapping genomic regions\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 19607  8021\n"
     ]
    }
   ],
   "source": [
    "multi.atac.ga.counts.list <- list()\n",
    "for (tissue in names(frag.files)) {\n",
    "    print(tissue)\n",
    "    selected.id <- which(batch == tissue)\n",
    "    selected.multi.count <- multi.atac.count[, selected.id]\n",
    "    colnames(selected.multi.count) <- gsub(barcode_suffix[tissue], '1', colnames(selected.multi.count))\n",
    "    selected.frag.path <- file.path(frag.path, frag.files[tissue])\n",
    "    selected.ga <- convert.gene.activtity(selected.multi.count, selected.frag.path, annotations, genome='hg38')\n",
    "    colnames(selected.ga) <- gsub('1', barcode_suffix[tissue], colnames(selected.ga))\n",
    "    print(dim(selected.ga))\n",
    "    multi.atac.ga.counts.list[[tissue]] <- selected.ga\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "34fac7af",
   "metadata": {},
   "outputs": [],
   "source": [
    "multi.atac.ga.counts <- do.call(cbind, multi.atac.ga.counts.list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "0df19ec5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Length1, Length2, Length intersection, Identical:\"\n",
      "[1] \"19607 13431 11141 FALSE\"\n"
     ]
    }
   ],
   "source": [
    "# Check features\n",
    "print.names.match(rownames(multi.atac.ga.counts), readLines('multiome/RNA/genes.txt'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "5a803eec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"Length1, Length2, Length intersection, Identical:\"\n",
      "[1] \"18422 18467 18422 FALSE\"\n"
     ]
    }
   ],
   "source": [
    "# Check cell barcodes\n",
    "print.names.match(colnames(multi.atac.ga.counts), colnames(multi.atac.count))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4df655e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset common cells and re-save multiome ATAC files\n",
    "path <- 'multiome/ATAC'\n",
    "write.counts.transpose(multi.atac.count[, colnames(multi.atac.ga.counts)], path, \n",
    "                       anno=multi.atac.anno[colnames(multi.atac.ga.counts)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "61d314aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset common cells and re-save multiome RNA files\n",
    "path <- 'multiome/RNA'\n",
    "multi.rna.data <- load.data.folder(path)\n",
    "multi.rna.count <- multi.rna.data$counts\n",
    "multi.rna.anno <- multi.rna.data$annotations\n",
    "write.counts.transpose(multi.rna.count[, colnames(multi.atac.ga.counts)], path, \n",
    "                       anno=multi.rna.anno[colnames(multi.atac.ga.counts)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "1d128b22",
   "metadata": {},
   "outputs": [],
   "source": [
    "path <- 'multiome/ATAC-gene-activities'\n",
    "write.counts.transpose(multi.atac.ga.counts, path, anno=multi.atac.anno[colnames(multi.atac.ga.counts)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b255245b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d31d7d02",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c29f08da",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4.0.5",
   "language": "R",
   "name": "ir405"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
